### ISSUE TITLE
Empty output on circular dependencies

### ISSUE THREAD

**Issue by enlait** [systemjs/builder#352]
After updating from 0.13.* to 0.14.* build began producing empty output without any error indications. 

I narrowed it down to circular dependencies. I use es6 modules and babel, config below

```
var builder = new Builder({
	baseURL     : "./src/main/js",
	transpiler  : 'babel',
	babelOptions: {
		modules  : 'es6',
		blacklist: [],
		comments : false
	}
});

builder
	.build('fail.js', 'fail-bundle.js', {minify: true, sourceMaps: true})
	.then(function () {
		console.log('Build complete');
	})
	.catch(function (err) {
		console.log('Build error');
		console.log(err);
	});
```

File contents:
fail.js:
```
import "fail2.js";
console.log("fail"); // just anything to have some code
```

fail2.js:
```
import "fail.js";
```

fail-bundle.js (output):
```

//# sourceMappingURL=fail-bundle.js.map
```

Also tried with export and default export, same result. Staying with 0.13 for now

**Comment by guybedford** [comment]
Odd, this works for me fine. If you add an `output` argument to your build, perhaps check the contents of `output.modules`, `output.trace` and `output.source` to see where the issue is? If the trace sources look ok, try removing your custom babelOptions?

**Comment by enlait** [comment]
I re-checked and I have this issue in 0.14.8 (and I did in 0.14.0 so probably anything in between too) but **0.14.9 seems to work fine**. But it wouldn't hurt to investigate to prevent regression.

I tried printing the whole output to console, here's what I get (0.14.8):
```
{ source: '',
  sourceMap: { add: [Function: add], get: [Function], toString: [Function] },
  modules: [ 'fail.js', 'fail2.js' ],
  entryPoints: [],
  tree:
   { 'fail.js':
      { name: 'fail.js',
        path: 'fail.js',
        metadata: [Object],
        deps: [Object],
        depMap: [Object],
        source: 'import "fail2.js";\r\nconsole.log("fail"); // just anything to have some code',
        runtimePlugin: null },
     'fail2.js':
      { name: 'fail2.js',
        path: 'fail2.js',
        metadata: [Object],
        deps: [Object],
        depMap: [Object],
        source: 'import "fail.js";',
        runtimePlugin: null } } }
```

I removed not just babelOptions, but everything except baseURL, as well as minification options, still the same. I should have mentioned I used babel-core 5.8.25 in all builds, which is latest on npm.

**Comment by enlait** [comment]
Update: I in fact still have this problem on 0.14.9 with prod build.

**Comment by guybedford** [comment]
This could be due to the missing entryPoints entry. Are you still getting the error when entryPoints is populated?

**Comment by enlait** [comment]
Huh? 
What are those entry points in the first place? Am I missing something in my config? I ran to docs after reading your comment, but I can't quite find anything about specifying entry points manually.

**Comment by guybedford** [comment]
@enlait in your previous comment you showed two examples - one with and one without entrypoints in the list. This array is generated internally - you don't need to specify it, but I just wanted to know if your failing case has entrypoints as an empty array? An empty entrypoints array would definitely mean the build would output nothing.

**Comment by enlait** [comment]
These two outputs are for the same minimalistic test case from the first comment, with 2 files importing each other (this is important part, without circular dependency everything's fine).
First output is for systemjs-builder 0.14.8 and second is for 0.14.9. As you can see, build fails in 0.14.8, and works for 0.14.9.

In my real project build also results in empty output using 0.14.9, but to be honest I didn't check if `entryPoints` array is empty - will do tomorrow. I suspect it is, since symptoms are the same.

Why 0.14.8 when 0.14.9 is out: if you didn't fix anything related to dependency resolution in 0.14.9, I speculate that it's the same bug, and it would be simpler to find it in 0.14.8, since it fails on just two files. Of course no sooner than you can reproduce the issue.

**Comment by enlait** [comment]
Played with test sources a bit, and have one more observation: build fails only if an entry point (that is, the js file from which I start making bundle via `builder.bundle` / `builder.build`) is part of dependency loop.

For example A.js imports B.js, B.js imports C.js, C.js imports A.js - `builder.build("A.js")` results in empty `entryPoints` . If B imports C and C imports B, but neither imports A then build succeeds.

This is true for 0.14.8.

**Comment by enlait** [comment]
Soo, disregard my doubts above, the bug is definitely there in 0.14.9 too.

Builder produces empty output on this input: https://gist.github.com/enlait/30266c703e551df8a627

The difference between 0.14.8 and 0.14.9 is that 0.14.8 fails on dependency loop of size 2, whereas 0.14.9 doesn't; but both fail with bigger loops.

**Comment by guybedford** [comment]
@enlait thanks. I assume `out.entryPoints.length == 0` in that case as well. The entry point detection does struggle with circular references - I will look into this.

**Comment by guybedford** [comment]
Thanks, fixed in https://github.com/systemjs/builder/commit/d82be245c53c631213c703ae350094d96ee0a76b.

**Comment by guybedford** [comment]
//cc @Munter the above commit changes the build order from post-order to reverse pre-order in order to handle the detection of circular entry points, but this should still provide the correct dependency execution order for your use cases.

**Comment by Munter** [comment]
Ok. I'll see if I run into any hardships when I get the build runtime up

**Comment by guybedford** [comment]
Released in 0.14.10.
